### PR TITLE
Extend Pro-Gen importer functionality.

### PR DESCRIPTION
**Pull request by apoettker**
﻿Beside some minor bug fixing I've improved the processing time due to dropping useless data while reading from the import file. Previously complete automatic logic I've extended the importer to an interactive tool with different options. Goal is extending the importer with different useful options.

screenshot pro-gen 170610

There are four blocks (Import Text, Tag Text, Primary Object and Option) organize the four different area to organize the import process. 
Import Text uses the default variables – if available – out of global settings. Independent from this data text's can be modified. While the Citation Page is a logical placeholder to store import informations, string directives like %d.%m.%Y will be substituted on the fly. The additional Citation Attribute is useful in times of merging similar data from same source, but different times (in between they maybe changed)! It is another reference to the import source. Citation Confidence is here identical used as in the main program. If necessary Source as well Citation data can be protected by Private key.
Tag Text has a Default area with Text, import File name and current Date (with standard Date Editor). The text entries are fully modifiable, the default text is a suggestion. There are seven rows contains the primary objects (Person, Family, Event, Place, Source, Citation, Note) with a Object button to activate / deactivate.  The (hidden) Text button copies the Default Text in the Tag Text entry. It's depend on the sensitivity controlled by the object access button at the beginning of the row. Event the (hidden) File and Date buttons copies the default values into the Tag Text area.
Within the Primary Object (Person, Family, Child) buttons import process can be tailored or extended. These keys activates the import of Persons, Families and Children. Default for all is On.
Miscellaneous Options can be switched in the last block. The Identifier (Person, Family) option use the original Pro-Gen identifiers as Gramps Id or generates new from scratch. The Date (Birth, Death) option set the birth resp. death time additional to the attribute entry as description in an event  (default: off). The Death Cause option (default: on) generates  an additional event with even this text as description. If this is switched off, the death cause will be stored in the description of the death event. REFN codes (default: off) are sometimes in the import files. They appear more with the original Pro-Gen program, but can be added as attribute to the person resp. family.

### CODE REVIEW

**Comment by Paul-Franklin** [comment]
Thanks for the addition to gramps!

I have not looked at your p.r. very closely but I see you added
two new files.  I assume they should be in MANIFEST.in also?

Also, as a new feature, it should be based on "master" (as
opposed to gramps50 or gramps42), as new features are
never put into a "maintenance" branch.

Do all unittests still pass?  (Run "python3 setup.py test" in
master to see.)

**Comment by apoettker** [comment]
Paul,

- MANIFEST.in is very short and general. What was the intention of this question?
- The extension is based on "master" yesterday noon. This extension for existing Pro-Gen importer is not intended to gramps50 or gramps42, just put it into the next release.
- All unittests passed for this extension.

Thanks a lot for the advise!

**Comment by macolquhoun** [comment]
> All unittests passed for this extension.

Ah no they did not pass on Travis see [here](https://travis-ci.org/gramps-project/gramps/builds/241663406?utm_source=github_status&utm_medium=notification)

**Comment by apoettker** [comment]
This is right there are bugs in the system. I've seen a similar output from "python3 setup.py test" as Paul suggested. I've got 2 errors ('jsonschema' and 'timeline-2.svg'), 93 failures and 56 skipped. So I'm not really sure where to dig out the failure related to this extension and to proceed the PR.

**Comment by prculley** [comment]
I suspect that the timeline-2.svg failure on your own system is cause by
NOT having a default environment.  In particular you probably have a page
size or orientation setting for output changed.  I usually rename or delete
the report_options.xml file in your profile's gramps directory.

I've sent you a private email regarding your import issue.

Paul C.

**Comment by Paul-Franklin** [comment]
Since you are still making changes, I suggest you add the
"work in progress" label to this p.r.

**Comment by apoettker** [comment]
Paul,

 

I'll like to do this. But it looks that I'm missing the rights. I did'nt found after reading all the online help the button to add/ remove a label.

 

Can you maybe send me a screenshot where I rather should look?

 

Thanks in advance!

Alois

 

Gesendet: Montag, 12. Juni 2017 um 18:18 Uhr
Von: "Paul Franklin" <notifications@github.com>
An: gramps-project/gramps <gramps@noreply.github.com>
Cc: "Alois Poettker" <alois.poettker@gmx.de>, Author <author@noreply.github.com>
Betreff: Re: [gramps-project/gramps] Extent Pro-Gen importer functionality. (#415)


Since you are still making changes, I suggest you add the
"work in progress" label to this p.r.

**Comment by Paul-Franklin** [comment]
One of the other developers already added it to this p.r.

**Comment by apoettker** [comment]
Paul Culley was so nice! But for future work I like to do it by myself, but I'm missing the rights.

@Nick-Hall: Would it possible to give the right to apply and delete labels and milestones for my PRs?

**Comment by Nick-Hall** [comment]
@apoettker I have sent you an invitation to join the Gramps Project organisation.

**Comment by Nick-Hall** [comment]
@apoettker I had to give you write [permissions](https://help.github.com/articles/repository-permission-levels-for-an-organization/) to apply labels and milestones.  We don't usually do this for new developers, so please continue to make pull requests for now.

File: gramps/plugins/lib/libprogen.py [hunk 1]
```diff
@@ -516,32 +515,28 @@ def parse_progen_file(self):
...
-            self.reset(_("Import from Pro-Gen"))
-            self.set_total(2 * len(self.pers) + len(self.rels))
+        # calculate total amount of data
+        if not self.uistate:
+            # approx. (1x father, 1x mother) + 1.5x child & families
+            self.set_total(2.5 * len(self.pers) + len(self.rels))
 
+        self.dbase.disable_signals()
         with DbTxn(_("Pro-Gen import"), self.dbase, batch=True) as self.trans:
-            self.dbase.disable_signals()
```
Comment (prculley) [inline gramps/plugins/lib/libprogen.py]: self.dbase.disable_signals() is not necessary if you use batch=True; the db does this for you.  (and the matching enable_signals)

File: gramps/plugins/lib/libprogen.py [hunk 2]
```diff
@@ -925,7 +914,13 @@ def __get_or_create_citation(self, source_title, date_text, page_text='', page_r
                 sattr.set_value(attr_text)
                 citation.add_attribute(sattr)
 
-        self.dbase.commit_citation(citation, self.trans)
+            self.dbase.add_citation(citation, self.trans)
+            self.dbase.commit_citation(citation, self.trans)
```
Comment (prculley) [inline gramps/plugins/lib/libprogen.py]: It is not necessary to commit after an add; the db add_citation does the commit_citation for you (after adding Gramps_ID and handle).  A bit better performance if you don't do both.

**Comment by Nick-Hall** [comment]
@apoettker The dialog is too large.  It should be designed for an 800x600 resolution screen.

I suggest you use notebook tabs to break the options into sections.  Use bold headings and indentation rather than frames.  For an example, see the [Gnome2 HIG](https://developer.gnome.org/hig-book/2.32/hig-book.html#design-window). Try to avoid using abbreviations unless really necessary.

Please run pylint on the code.  New files should aim to have a pylint score of 9.0 or greater. See our [programming guidelines](https://gramps-project.org/wiki/index.php?title=Programming_guidelines#Pylint).  Improving the pylint score of existing code is also very much appreciated in _master_.  Please put changes that tidy up existing code in a separate commit.

File: gramps/plugins/lib/libprogen.py [hunk 3]
```diff
@@ -0,0 +1,1850 @@
+# -*- coding: utf-8 -*-
+#
+# Gramps - a GTK+/GNOME based genealogy program
+#
+# Copyright (C) 2008-2017  Kees Bakker (sen. / jun.)
```
Comment (Nick-Hall) [inline gramps/plugins/lib/libprogen.py]: The copyright for Kees Bakker appears to have been changed.  Please check.

**Comment by apoettker** [comment]
@Nick-Hall: dialogue new arranged and slightly improved (see screen shots below):

screenshot pro-gen i 170624
screenshot pro-gen ii 170624
screenshot pro-gen iii 170624

PyLint level is now 9.2 for importprogen.py (w/o classes Glade, Buildable) and libprogen.py

**Pull request merged**